components/ai_chat/core/browser/conversation_handler.cc

```diff
@@ -0,0 +1,1408 @@
+  for (std::vector<mojom::ConversationTurnPtr>::reverse_iterator rit =
+           chat_history_.rbegin();
+       rit != chat_history_.rend(); ++rit) {
+    if (rit->get()->character_type == CharacterType::HUMAN) {
+      auto turn = *std::make_move_iterator(rit);
+      auto human_turn_iter = rit.base() - 1;
+      chat_history_.erase(human_turn_iter, chat_history_.end());
+      SubmitHumanConversationEntry(std::move(turn));
+      break;
+    }
+  }
+}
+
+void ConversationHandler::GetAPIResponseError(
+    GetAPIResponseErrorCallback callback) {
+  std::move(callback).Run(current_error_);
+}
+
+void ConversationHandler::ClearErrorAndGetFailedMessage(
+    ClearErrorAndGetFailedMessageCallback callback) {
+  DCHECK(!chat_history_.empty());
+
+  SetAPIError(mojom::APIError::None);
+  mojom::ConversationTurnPtr turn = std::move(*chat_history_.end());
+  chat_history_.erase(chat_history_.end());
+
+  OnHistoryUpdate();
+
+  std::move(callback).Run(std::move(turn));
+}
+
+void ConversationHandler::SubmitSelectedText(
+    const std::string& selected_text,
+    mojom::ActionType action_type,
+    GeneratedTextCallback received_callback,
+    EngineConsumer::GenerationCompletedCallback completed_callback) {
+  const std::string& question = GetActionTypeQuestion(action_type);
+  SubmitSelectedTextWithQuestion(selected_text, question, action_type,
+                                 std::move(received_callback),
+                                 std::move(completed_callback));
+}
+
+void ConversationHandler::SubmitSelectedTextWithQuestion(
+    const std::string& selected_text,
+    const std::string& question,
+    mojom::ActionType action_type,
+    GeneratedTextCallback received_callback,
+    EngineConsumer::GenerationCompletedCallback completed_callback) {
+  if (received_callback && completed_callback) {
+    // Start a one-off request and replace in-place with the result.
+    // TODO(petemill): This should only belong in the caller location,
+    // such as ai rewriter dialog (or a shared utility).
+    engine_->GenerateRewriteSuggestion(
+        selected_text, question,
+        base::BindRepeating(
+            [](GeneratedTextCallback received_callback,
+               mojom::ConversationEntryEventPtr rewrite_event) {
+              constexpr char kResponseTagPattern[] =
+                  "<\\/?(response|respons|respon|respo|resp|res|re|r)?$";
+              if (!rewrite_event->is_completion_event()) {
+                return;
+              }
+
+              std::string suggestion =
+                  rewrite_event->get_completion_event()->completion;
+
+              base::TrimWhitespaceASCII(suggestion, base::TRIM_ALL,
+                                        &suggestion);
+              if (suggestion.empty()) {
+                return;
+              }
+
+              // Avoid showing the ending tag.
+              if (RE2::PartialMatch(suggestion, kResponseTagPattern)) {
+                return;
+              }
+
```

Comment:
This code block is copy-pasted in ai_rewriter_ui.cc, please extract it into a util function in brave/components/ai_chat/core/browser/utils.h/cc to be shared.

Reply:
I had left it because I know we're going to remove it at some point - this shouldn't be done in this layer, it is engine-dependent and most engines do not need it since they strip it out with stop sequences. But I have made a util function in utils.cc - `BindParseRewriteReceivedData` which the rewriter dialog and in-line rewriter use. In the future we'll likely remove the in-line rewriter in favor of the rewriter dialog. I also removed all this from ConversationHandler as it does not need to be done here at all - nothing is calling it.